feat: add two new methods to search for products, search_iter_page and search_all

maximlt opened the pull request:
Fixes #191 

Two new methods for `EODataAccessGateWay`:
* `search_iter_page`: to iterate over the pages of a products search
* `search_all`: to get all the products matching some search criteria -> it uses `seach_iter_page` internally to iterate over all the pages that return products

`search_all` tries to be a little smarter than `search_iter_page` or  `search`, in the sense that it tries to minimize the number of requests made. It makes use of an internal parameter `max_items_per_page` configured per provider (see `providers.yml`). 8 providers have such a parameter configured, I have found these values either in their documentation (and checked it) or by trial and error. If this internal parameter isn't set, a default value of 50 is used. I chose it because it's the lowest one between all the providers (mundi) but it's still higher than the default items_per_page of 20 used by `search` originally. The user can also defined the value she wants to use dynamically (`search_all(items_per_page=...)`), which could come in handy if the value we have set doesn't work properly.

Note also that `search_iter_page` stops yielding products if the number of products found is less than the number of items requested. This spares one last request.

Next to that:
* The logic that prepared the search kwargs passed to `search_plugin.query` was extracted from `search` and put into a new private method `_prepare_search`. It's used in the three search methods.
* `test_core.py` was refactored to make it easier to add tests
* Lots of unit tests were added to `test_core.py`, for: `guess_by_product`, `_prepare_search`, `_do_search`, `search_iter_page` and `search_all`
* An end-to-end test was added for searching a product by ID (just a provider tested, sobloo)
* There is no a log warning emitted when `box` or `bbox` are used in any of the search methods. It recommends using `geom` instead

As discussed privately, there appears to be some search plugins or providers that do not support pagination. This is a problem for `search_iter_page` - and `search_all` as a consequence - since it means it would just iterate indefinitely. A solution to this problem was implemented, by comparing the first product of the next page with the first product of the previous one (their `properties["id"]` and `provider` values). The iteration stops if they are equal. There is a test for this approach, which I have also tested manually with `astraea_eod` whose API only implements a *limit* and not a *page*.

I plan to add a proper documentation to these two new methods in a following PR, since the tutorials need to be refactored a little.

sbrunato commented:
For `astraea_eod`, STAC API pagination only needs `item["links"][]["rel"] == "next"`. `page` parameter is only optional, see https://github.com/radiantearth/stac-api-spec/tree/master/item-search#paging.

For STAC providers, we should this `next` entry by default instead of `page`+`limit`.

This could be done by adding a new parameter in `providers.yml` / `stac_provider.yml`:
```yml
  pagination:
    next_page_url_key_path: '$.links[?(@.rel="next")].href'
```
And use `jsonpath_ng.ext.parse` instead of `jsonpath_ng.parse`, see https://github.com/h2non/jsonpath-ng/issues/8

maximlt commented:
@sbrunato I have implemented a mechanism to retrieve the next page when a provider doesn't implement "page".

In `search_iter_page` a search plugin is introspected to check whether it has a `pagination.next_page_url_key_path` key. If it does, it is assumed that it doesn't implement "page", and a kwarg `set_next_page_url` is passed down to `.query`. This is handled within `QueryStringSearch.do_search` where the response (JSON only right now) is parsed to get the "next" URL. It replaces the configuration key `pagination.next_page_url_tpl`. So the next call to `.query` will use that URL and will update it to get the next page.

I have added comments since this logic is quite special. Maybe we'll come up with a better way in the future.

maximlt commented:
@sbrunato I've made the changes as we agreed:
* `search_iter_page` doesn't propagate a boolean to trigger to "next" search through the plugin infrastructure down to `do_search`. Instead `do_search` takes care of setting an attribute `next_page_url` if the pagination config has the key `next_page_url_key_path`.
* `search_iter_page` uses that URL after the first iteration if available. It uses it by replacing `search_plugin.config.pagination["next_page_url_tpl"]`
* `search_plugin.config.pagination["next_page_url_tpl"]` and `search_plugin.next_page_url` are both reset after each iteration. The generator holds the `next_page_url` value in its internal state. This has the advantage that the search plugin is in a clean state after `next(dag.search_iter_page(**search_criteria))`. I thought it was quite important that `search_plugin.next_page_url` was reset every time, I didn't want a URL from a previous search to mess up with a new search (different criteria).

sbrunato commented:
updates #133
